packages/typescript/src/internal/ts_project.bzl

```diff
@@ -0,0 +1,273 @@
+"ts_project rule"
+
+load("@build_bazel_rules_nodejs//:providers.bzl", "DeclarationInfo", "NpmPackageInfo", "run_node")
+
+_ATTRS = {
+    # NB: no restriction on extensions here, because tsc sometimes adds type-check support
+    # for more file kinds (like require('some.json')) and also
+    # if you swap out the `compiler` attribute (like with ngtsc)
+    # that compiler might allow more sources than tsc does.
+    "srcs": attr.label_list(allow_files = True, mandatory = True),
+    "extends": attr.label_list(allow_files = [".json"]),
+    "tsc": attr.label(default = Label("@npm//typescript/bin:tsc"), executable = True, cfg = "host"),
+    "tsconfig": attr.label(mandatory = True, allow_single_file = [".json"]),
+    "deps": attr.label_list(providers = [DeclarationInfo]),
+}
+
+# tsc knows how to produce the following kinds of output files.
+# NB: the macro `ts_project_macro` will set these outputs based on user
+# telling us which settings are enabled in the tsconfig for this project.
+_OUTPUTS = {
+    "buildinfo_out": attr.output(),
+    "js_outs": attr.output_list(),
+    "map_outs": attr.output_list(),
+    "typing_maps_outs": attr.output_list(),
+    "typings_outs": attr.output_list(),
+}
+
+_TsConfigInfo = provider(
+    doc = """Passes tsconfig.json files to downstream compilations so that TypeScript can read them.
+        This is needed to support Project References""",
+    fields = {
+        "tsconfigs": "depset of tsconfig.json files",
+    },
+)
+
+def _ts_project_impl(ctx):
+    arguments = ctx.actions.args()
+    arguments.add_all([
+        "-p",
+        ctx.file.tsconfig.short_path,
+        "--outDir",
+        "/".join([ctx.bin_dir.path, ctx.label.package]),
+    ])
+    if len(ctx.outputs.typings_outs) > 0:
+        arguments.add_all([
+            "--declarationDir",
+            "/".join([ctx.bin_dir.path, ctx.label.package]),
+        ])
+
+    # When users report problems, we can ask them to re-build with
+    # --define=VERBOSE_LOGS=1
+    # so anything that's useful to diagnose rule failures belongs here
+    if "VERBOSE_LOGS" in ctx.var.keys():
+        arguments.add_all([
+            # What files were in the ts.Program
+            "--listFiles",
+            # Did tsc write all outputs to the place we expect to find them?
+            "--listEmittedFiles",
+            # Why did module resolution fail?
+            "--traceResolution",
+            # Why was the build slow?
+            "--diagnostics",
+            "--extendedDiagnostics",
+        ])
+
+    deps_depsets = []
+    for dep in ctx.attr.deps:
+        if _TsConfigInfo in dep:
+            deps_depsets.append(dep[_TsConfigInfo].tsconfigs)
+        if NpmPackageInfo in dep:
+            # TODO: we could maybe filter these to be tsconfig.json or *.d.ts only
+            # we don't expect tsc wants to read any other files from npm packages.
+            deps_depsets.append(dep[NpmPackageInfo].sources)
+        elif DeclarationInfo in dep:
```

Comment:
should this just be an `if`? They seem orthogonal and should overlap anyway.

Reply:
hmm, I think this is true. Need to verify if we have coverage for it